bpo-5680: IDLE: Customize running a module.

csabella opened the pull request:
The initialize options are 1) add command line options, which are appended to sys.argv as if passed on a real command line, and 2) skip the shell restart.  The customization dialog is accessed by a new item on the Run menu.
<!-- issue-number: [bpo-5680](https://bugs.python.org/issue5680) -->
https://bugs.python.org/issue5680
<!-- /issue-number -->

csabella committed:
bpo-5680: IDLE: Run modules with command line arguments

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

csabella committed:
Fix editing error

csabella committed:
Create more generic dialog and change keybinding

csabella commented on Lib/idlelib/query.py:
```diff
@@ -83,7 +84,7 @@ def __init__(self, parent, title, message, *, text0='', used_names={},
             self.wait_window()
 
-    def create_widgets(self):  # Call from override, if any.
+    def create_widgets(self, ok_text='OK'):  # Call from override, if any.
```
I made this configurable so that the 'Custom Run' dialog could change the `OK` to `Run`.

terryjreedy replied:
Nice alternative, at least for now, to factoring out the query code I want to reuse and starting with a separate class.  There is another issue about standardizing dialog box behavior that we should revisit after this one.

csabella replied:
Sounds good!  🙂

csabella commented on Lib/idlelib/query.py:
```diff
@@ -302,10 +303,61 @@ def entry_ok(self):
+
+        restart.grid(column=0, row=4, columnspan=3, padx=5, sticky='w')
+        self.args_error.grid(column=0, row=12, columnspan=3, padx=5,
+                             sticky='we')
```
I'm not using `self.args_error` yet, but it added nice spacing and I figured it would come in handy for an edit such as `must restart with arguments` as mentioned on the bug tracker.

terryjreedy replied:
I will have to reread more of the query code to understand that.  We will have to comment out or remove the restart button until it works.

csabella replied:
Yes, `restart` works now, but it's never disabled.  I wasn't sure which edits would make sense around it.  On the ticket, you had mentioned that using CLI arguments would make sense on a restart, but right now it allows changing the CLI arguments without restarting the shell.  That's where my other comments (about this dialog not allowing blank arguments) and what the current `Run Module`'s behavior should be.

terryjreedy replied:
I believe I covered this elsewhere.

csabella commented on Lib/idlelib/query.py:
```diff
@@ -302,10 +303,61 @@ def entry_ok(self):
+        cli_args = self.entry.get().strip()
+        if not cli_args:
+            self.showerror('no arguments specified.')
+            return None
```
Not sure about this edit, but it makes everything a little nicer right now.  The reason I'm not sure about it is due to the behavior of the current `Run Module`.  Should `Run Module` use the last settings from `Custom Run` or should it work exactly as it works now (while in a subprocess)?  `Run Module` without a subprocess uses the `Custom Run` settings from the last run.

taleinat replied:
IMO "Run Module" should always ignore the latest "Custom Run" settings, regardless of whether or not a subprocess is used.

terryjreedy replied:
I believe Run Module is currently unchanged and agree that this should continue.

csabella commented on Lib/idlelib/query.py:
```diff
@@ -302,10 +303,61 @@ def entry_ok(self):
+        return lex
+
+    def restart_ok(self):
+        return self.restartvar.get()
```
Could add edits here for when `restart` cannot be False.

terryjreedy replied:
No longer possible.

csabella commented on Lib/idlelib/runscript.py:
```diff
@@ -131,22 +138,27 @@ def _run_module_event(self, event):
-                _basename(_sys.argv[0]) != _basename(__file__)):
-                _sys.argv = [__file__]
+                _basename(_sys.argv[0]) != _basename(__file__) or
+                len(argv) > 1):
```
Checking the length so that, if there are additional args, they will replace what was already there.  However, existing args won't be replaced by no args.  Currently, F5 still runs with no args and restarts the shell, but that's something to keep in mind if F5 changes.

terryjreedy replied:
I believe that this can only matter if one used the deprecated -n option.  I have no intention of changing F5

terryjreedy reviewed (changes requested):
Test of behavior without looking much at code yet.  Will respond to other comments later.  Immediate issue:  I have a custom keyset without new pseudoevent binding and on startup see

 Warning: config.py - IdleConf.GetCoreKeys -
 problem retrieving key binding for event '<<run-custom>>'
 from key set 'test'.
 returning default value: ['<Shift-Key-F5>']

The fix is simple and I know how to do it if you don't beat me. 

Separate issue if not done yet: write back keyset with new bindings.

Otherwise, works great.

csabella committed:
Don't show warning message for new key binding

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -242,6 +242,57 @@ def test_entry_ok_helpsource(self):
+    def test_blank_args(self):
+        dialog = self.Dummy_CustomRun(' ')
+        self.assertEqual(dialog.cli_args_ok(), None)
+        self.assertIn('no arguments', dialog.entry_error['text'])
```
This should use `assertRegex` to do a case-insensitive check.
```suggestion
        self.assertRegex(r'(?i)no arguments', dialog.entry_error['text'])
```

terryjreedy replied:
This assert is gone.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -242,6 +242,57 @@ def test_entry_ok_helpsource(self):
+    def test_invalid_args(self):
+        dialog = self.Dummy_CustomRun("'no-closing-quote")
+        self.assertEqual(dialog.cli_args_ok(), None)
+        self.assertIn('No closing', dialog.entry_error['text'])
```
This should use assertRegex to do a case-insensitive check.
```suggestion
        self.assertRegex(r'(?i)no closing', dialog.entry_error['text'])
```

terryjreedy replied:
The test is clear and it works.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -242,6 +242,57 @@ def test_entry_ok_helpsource(self):
+
+    class Dummy_CustomRun:
+        cli_args_ok = query.CustomRun.cli_args_ok  # Function being tested.
+        entry = Var()
```
Why is this a class attribute? Besides being unclear, this also doesn't reflect the actual `CustomRun` class.

terryjreedy replied:
I wrote the query tests, and Cheryl followed the pattern I set.  For testing with one instance at a time, or ever, class and instance attributes work the same.  For attributes never rebound, I attached them to the class, especially if there was no other need for an __init__ method.  Very clear to me.  However, I moved Var attributes to __init__ when there is one as it saves a line.

The cli_args_ok method being tested accesses self.query.  tk Vars and Entries both have a get method that does the same thing, so a mock Var replaces an Entry quite nicely.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -242,6 +242,57 @@ def test_entry_ok_helpsource(self):
+    class Dummy_CustomRun:
+        cli_args_ok = query.CustomRun.cli_args_ok  # Function being tested.
+        entry = Var()
+        entry_error = {}
```
It might be simpler to make `Dummy_CustomRun` a `Mock` instance and then check calls to `showerror()`, rather than the `entry_error['text']` stuff which is more implementation-specific.

terryjreedy replied:
I used this in all the other functional method tests.  The dummy classes have and do exactly what is needed to test the method.  A unittest.mock would be much more complicated and difficult and would run much slower.  Test speed is important.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -242,6 +242,57 @@ def test_entry_ok_helpsource(self):
+        for dialog.restart in {True, False}:
+            for cli_args, result in ((None, None),
+                                     ('my arg', ('my arg', dialog.restart))):
+                with self.subTest():
```
```suggestion
                with self.subTest(restart=dialog.restart, cli_args=cli_args):
```

terryjreedy replied:
Done.

taleinat commented on Lib/idlelib/runscript.py:
```diff
@@ -113,15 +114,21 @@ def run_module_event(self, event):
+            return 'break'
+        return self._run_module_event(event, cli_args=run_args[0], restart=run_args[1])
+
+    def _run_module_event(self, event, *, cli_args=None, restart=True):
```
The "Run Module" and "Run Custom" should each have their own (simple) event handler, with the common code extracted into a common function used by both, which isn't an event handler.

terryjreedy replied:
_run_module_event, with initial '_', is already the common code for run_module_event and run_custom_event.  It is only an 'event handler' to implement the Mac-specific delayed call via an internal pseudo pseudoevent.  I am assuming that someone tried calling it directly after the delay.  As a separate issue, we might test whether the workaround is needed in 8.6.

taleinat commented on Misc/NEWS.d/next/IDLE/2019-06-03-00-39-29.bpo-5680.VCQfOO.rst:
```diff
@@ -0,0 +1 @@
+Run modules from the editor using command line arguments.
```
This needs to be a bit more explicit, mentioning that a new menu option is added.

terryjreedy replied:
Expanded.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -348,5 +399,23 @@ def test_click_help_source(self):
+        dialog =  query.CustomRun(root, 'Title', _utest=True)
+        dialog.entry.insert(0, 'okay')
+        dialog.button_ok.invoke()
+        self.assertEqual(dialog.result, (['okay'], True))
```
The should probably be a call to `root.update()` after the `.invoke()` call, to ensure that Tk has a chance to do all of its processing.

terryjreedy replied:
I believe invoke blocks until done.  I have run this test at least 20 times without issue.  event_generate is different (and flakey).  There are 55 invokes in idle tests and my spot check of nearly 10 showed none followed by update.  And I know of no resulting failures.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -348,5 +399,23 @@ def test_click_help_source(self):
+        dialog.button_ok.invoke()
+        self.assertEqual(dialog.result, (['okay'], True))
+        del dialog
+        root.destroy()
```
I suggest cleaning up using `self.addCleanup()` immediately after creating the `Tk()` instance, e.g.:

```python
    root = Tk()
    def cleanup():
        root.update()
        root.destroy()
    self.addCleanup(cleanup)

terryjreedy replied:
I far prefer explicit calls.  Update is not generally needed.  update_idletasks is more often needed.  I have not seen any of the tclerror messages that so indicate.

taleinat commented on Lib/idlelib/idle_test/test_query.py:
```diff
@@ -348,5 +399,23 @@ def test_click_help_source(self):
+        dialog.entry.insert(0, 'okay')
+        dialog.button_ok.invoke()
+        self.assertEqual(dialog.result, (['okay'], True))
+        del dialog
```
Are this `del` call, and the one below, actually needed?

terryjreedy replied:
No.  Gone.  Added class attributes need deletion, but function locals already go on return, and there is no evident issue here with destroying first.

taleinat commented on Lib/idlelib/query.py:
```diff
@@ -302,10 +303,61 @@ def entry_ok(self):
+            self.showerror('no arguments specified.')
+            return None
+        try:
+            lex = shlex.split(cli_args, posix=True)
```
What does `lex` stand for? For me, this is confusing both here and in `entry_ok()`. Perhaps `args_list`?

terryjreedy replied:
Agreed. I switched to cli_string and cli_args.  cli_args_ok should return cli_args.

taleinat reviewed (commented):
As a general note, I'm not sure about the usefulness of the non-GUI tests for the RunCustom dialog (`CustomRunCLIargsokTest`, `CustomRunEntryokTest`) . They rely rather heavily on the internal implementation. Since we have GUI tests run by the CI now, I prefer to have simpler tests actually using Tk with less mocking. Would it be possible to simply move the tests from these classes into `CustomRunGuiTest`?

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_13763

terryjreedy committed:
Change menu entry 'Run Custom' to 'Run ... Customized'.
Revise corresponding doc entry.

terryjreedy committed:
Expand blurb.

terryjreedy committed:
Polish run.py.

terryjreedy commented on Lib/idlelib/runscript.py:
```diff
@@ -113,15 +114,21 @@ def run_module_event(self, event):
+        # User cancelled.
+        if not run_args:
+            return 'break'
+        return self._run_module_event(event, cli_args=run_args[0], restart=run_args[1])
```
Changed to pass run_args and unpack in _run_module.

terryjreedy committed:
In runscript, pass run_args.

terryjreedy committed:
More runscript edit.

terryjreedy committed:
Edit query and tests.

terryjreedy committed:
Move some test dummy attributes to instance.

terryjreedy committed:
A couple more.

terryjreedy committed:
Update query coverage.

terryjreedy committed:
pass rub_args by keyword

terryjreedy committed:
Merge branch 'master' into bpo5680

terryjreedy committed:
Show customize box after check, attach to shell.
Don't ask for config if still may toss.
Want focus to go to shell unless cancel.

terryjreedy committed:
Minor tweaks.
